auction-app-backend/src/main/java/com/example/auctionapp/service/implementation/ProductServiceImpl.java

```diff
@@ -29,11 +34,65 @@ public ProductServiceImpl(final ProductRepository productRepository, final Categ
     }
 
     @Override
-    public Page<Product> getProducts(final UUID categoryId, final String searchProduct, final int page, final int size) {
+    public Map<String, Object> getProducts(UUID categoryId, String searchProduct, int page, int size) {
         Pageable pageable = PageRequest.of(page, size);
         Specification<ProductEntity> specification = ProductSpecification.withDynamicQuery(categoryId, searchProduct);
 
-        return productRepository.findAll(specification, pageable).map(ProductEntity::toDomainModel);
+        Page<Product> products = productRepository.findAll(specification, pageable).map(ProductEntity::toDomainModel);
+        Map<String, Object> response = new HashMap<>();
+
+        response.put("products", products);
+
+        if (products.getTotalElements() < 1 && searchProduct != null && !searchProduct.isBlank()) {
+            String suggestedQuery = suggestCorrection(searchProduct);
+
+            // suggestion only if it is meaningfully different from the search query
+            if (suggestedQuery != null && !suggestedQuery.equalsIgnoreCase(searchProduct)) {
+                response.put("suggestion", suggestedQuery);
+            }
+        }
+
+        return response;
+    }
+
+    private String suggestCorrection(String query) {
```

Comment:
Can we move this to the util class, to avoid having custom logic for calculating the distance as part of the service.